Overhaul training "Mastering Plone 6 development" (#852)
plone/training: docs/mastering-plone-5/instructions_plone5/index.md

---
myst:
  html_meta:
    "description": "Set up environment for development"
    "property=og:description": "Set up environment for development"
    "property=og:title": "Installing Plone for the Training"
    "keywords": "Plone 5"
---

(plone5-instructions-label)=

# Installing Plone for the Training

Keep in mind that you need a fast Internet connection during installation since you'll have to download a lot of data!

> (plone5-instructions-no-vagrant-label)=

Reviewer: This should be moved downward to be just above the heading.

Author: done. thanks for the hint!

```

(plone5-instructions-vagrant-label)=

## Installing Plone with Vagrant

We use a virtual machine (Ubuntu 18.04) to run Plone during the training.

We rely on [Vagrant](https://www.vagrantup.com) and [VirtualBox](https://www.virtualbox.org) to give the same development environment to everyone.

[Vagrant](https://www.vagrantup.com) is a tool for building complete development environments.

We use it together with Oracle’s [VirtualBox](https://www.virtualbox.org) to create and manage a virtual environment.

(plone5-install-virtualbox)=

### Install VirtualBox

Vagrant uses Oracle’s VirtualBox to create virtual environments.

Here is a link directly to the download page: <https://www.virtualbox.org/wiki/Downloads>.

We use VirtualBox 6.0.x

(plone5-instructions-configure-vagrant-label)=

### Install and configure Vagrant

Get the latest version from <https://developer.hashicorp.com/vagrant/install> for your operating system and install it.

Now your system has a command {command}`vagrant` that you can run in the terminal.

First, create a directory in which you want to do the training.

```{warning}

> If you already have a {file}`training` directory because you followed the **Installing Plone without vagrant** instructions above,

Reviewer: ```suggestion
If you already have a {file}`training` directory because you followed the {ref}`plone5-instructions-no-vagrant-label` instructions above,
```

you should either delete it, rename it, or use a different name below.
```

```shell
mkdir training
cd training
```

Setup Vagrant to automatically install the current guest additions.
You can choose to skip this step if you encounter any problems with it.

```shell
vagrant plugin install vagrant-vbguest
```

Now download {download}`plone_training_config.zip <../../_static/plone_training_config.zip>` and copy its contents into your training directory.

```shell
wget https://github.com/plone/training/raw/refs/heads/main/docs/_static/plone_training_config.zip
unzip plone_training_config.zip
```

The training directory should now hold the file {file}`Vagrantfile` and the directory {file}`manifests` which again contains several files.

Now start setting up the virtual machine (VM) that is configured in {file}`Vagrantfile`:

```shell
vagrant up
```

> This takes a **veeeeery loooong time** (between 10 minutes and 1h depending on your Internet connection and system speed) since it does all the following steps:

Reviewer: ```suggestion
This takes a **veeeeery loooong time** (between 10 minutes and 1 hour depending on your Internet connection and system speed) since it does all the following steps:
```

- downloads a virtual machine (Official Ubuntu Server 18.04 LTS, also called "Bionic Beaver")
- sets up the VM
- updates the VM
- installs various system-packages needed for Plone development

> - clones the training buildout into /vagrant/buildout

Reviewer: ```suggestion
- clones the training buildout into `/vagrant/buildout`
```

- builds Plone annd installs all dependencies

````{note}
Sometimes this stops with the message:

```shell
Skipping because of failed dependencies
```
````

If this happens or you have the feeling that something has gone wrong and the installation has not finished correctly for some reason
you need to run the following command to repeat the process.

This will only repeat steps that have not finished correctly.

```shell
vagrant provision
```

> You can do this multiple times to fix problems, e.g. if your network connection was down and steps could not finish because of this.

Reviewer: ```suggestion
You can do this multiple times to fix problems, for example, if your network connection was down and steps could not finish because of this.
```

2019-03-07 10:38:21,160 WARNI [plone.behavior:172][MainThread] Specifying 'for' in behavior 'Tiles' if no 'factory' is given has no effect and is superfluous.
2019-03-07 10:38:22,473 WARNI [PrintingMailHost:30][MainThread] Hold on to your hats folks, I'm a-patchin'
2019-03-07 10:38:22,474 WARNI [PrintingMailHost:124][MainThread]

******************************************************************************

Monkey patching MailHosts to print e-mails to the terminal.

This is instead of sending them.

NO MAIL WILL BE SENT FROM ZOPE AT ALL!

Turn off debug mode or remove Products.PrintingMailHost from the eggs
or remove ENABLE_PRINTING_MAILHOST from the environment variables to
return to normal e-mail sending.

See https://pypi.python.org/pypi/Products.PrintingMailHost

******************************************************************************

2019-03-07 10:38:22,510 INFO  [Zope:44][MainThread] Ready to handle requests
Starting server in PID 25230.
Serving on http://0.0.0.0:8080
```

````{note}
In rare cases when you are using macOS with an UTF-8 character set starting Plone might fail with the following error:

```text
ValueError: unknown locale: UTF-8
```
````

> In that case you have to put the localized keyboard and language settings in the .bash_profile

Reviewer: ```suggestion
In that case you have to put the localized keyboard and language settings in the {file}`.bash_profile`
```

> of the vagrant user to your locale (like `en_US.UTF-8` or `de_DE.UTF-8`)

Reviewer: ```suggestion
of the Vagrant user to your locale (like `en_US.UTF-8` or `de_DE.UTF-8`).
```

```shell
export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8
```

Now the Zope instance we're using is running.
You can stop the running instance anytime using {kbd}`ctrl + c`.

If it doesn't, don't worry, your shell isn't blocked.

> Type {kbd}`reset` (even if you can't see the prompt) and press RETURN, and it should become visible again.

Reviewer: ```suggestion
Type {kbd}`reset` (even if you can't see the prompt) and press {kbd}`return`, and it should become visible again.
```

If you point your local browser at <http://localhost:8080> you see that Plone is running in Vagrant.

> This works because VirtualBox forwards the port 8080 from the guest system (the vagrant Ubuntu) to the host system (your normal operating system).

Reviewer: ```suggestion
This works because VirtualBox forwards the port 8080 from the guest system (the Vagrant Ubuntu) to the host system (your normal operating system).
```

> There is no Plone site yet - we will create one in chapter 6.

Reviewer: ```suggestion
There is no Plone site yet - we will create one in {doc}`/mastering-plone-5/features`.
```

The Buildout for this Plone is in a shared folder.

> This means we run it in the vagrant box from {file}`/vagrant/buildout` but we can also access it in our own operating system and use our favorite editor.

Reviewer: ```suggestion
This means we run it in the Vagrant box from {file}`/vagrant/buildout` but we can also access it in our own operating system and use our favorite editor.
```

You will find the directory {file}`buildout` in the directory {file}`training` that you created in the beginning
next to {file}`Vagrantfile` and {file}`manifests`.

```{note}

> The database and the python packages are not accessible in your own system since large files cannot make use of symlinks in shared folders.

Reviewer: ```suggestion
The database and the Python packages are not accessible in your own system since large files cannot make use of symlinks in shared folders.
```

> The database lies in `/home/vagrant/var`, the python packages are in `/home/vagrant/packages`.

Reviewer: ```suggestion
The database lies in `/home/vagrant/var`, the Python packages are in `/home/vagrant/packages`.
```

```

If you have any problems or questions please mail us at <mailto:team@starzel.de> or create a ticket at <https://github.com/plone/training/issues>.

(plone5-instructions-vagrant-does-label)=

### What Vagrant does

> Installation is done automatically by vagrant and puppet.

Reviewer: ```suggestion
Installation is done automatically by Vagrant and Puppet.
```

If you want to know which steps are actually done please see the chapter {doc}`what_vagrant_does`.

(plone5-instructions-vagrant-care-handling-label)=

```{note}
**Vagrant Care and Handling**

Keep in mind the following recommendations for using your Vagrant VirtualBoxes:

- Use the {command}`vagrant suspend` or {command}`vagrant halt` commands to put the VirtualBox to "sleep" or to "power it off" before attempting to start another Plone instance anywhere else on your machine, if it uses the same port.  That's because vagrant "reserves" port 8080, and even if you stopped Plone in vagrant, that port is still in use by the guest OS.

> - If you are done with a vagrant box, and want to delete it, always remember to run {command}`vagrant destroy` on it before actually deleting the directory containing it.  Otherwise you'll leave its "ghost" in the list of boxes managed by vagrant and possibly taking up disk space on your machine.

Reviewer: ```suggestion
- If you are done with a vagrant box, and want to delete it, always remember to run {command}`vagrant destroy` on it before actually deleting the directory containing it.  Otherwise you'll leave its "ghost" in the list of boxes managed by Vagrant and possibly taking up disk space on your machine.
```
